A81-xds-authority-rewriting.md

```diff
@@ -0,0 +1,171 @@
+* [gRFC A29: xDS mTLS Security][A29]
+* [gRFC A30: xDS v3 Support][A30]
+* [gRFC A31: xDS Timeout Support and Config Selector][A31]
+* [gRFC A37: xDS Aggregate and Logical DNS Clusters][A37]
+* [gRFC A60: xDS-Based Stateful Session Affinity for Weighted Clusters][A60]
+* [gRFC A14: Channelz][A14]
+
+[A27]: A27-xds-global-load-balancing.md
+[A28]: A28-xds-traffic-splitting-and-routing.md
+[A29]: A29-xds-tls-security.md
+[A30]: A30-xds-v3.md
+[A31]: A31-xds-timeout-support-and-config-selector.md
+[A37]: A37-xds-aggregate-and-logical-dns-clusters.md
+[A60]: A60-xds-stateful-session-affinity-weighted-clusters.md
+[A14]: A14-channelz.md
+
+## Proposal
+
+This proposal has several parts:
+- Bootstrap config change: We will add a new server feature in the bootstrap
+  config to indicate that we should allow authority rewriting.
+- xDS resource validation: We will process authority rewriting fields in RDS
+  resources and the hostname field in EDS resources, and return that
+  information to the XdsClient watchers as part of the parsed resources.
+- xDS ConfigSelector changes: We will change the xDS ConfigSelector to enable
+  authority rewriting based on the chosen route.
+- xds_cluster_impl LB policy changes: We will change the xds_cluster_impl LB
+  policy to actually perform the authority rewriting if it was enabled by the
+  ConfigSelector.
+
+### Server Feature in Bootstrap Config
+
+In order to address use-cases where authority rewriting may not be
+acceptable from a security perspective, we will add a new server feature
+to the bootstrap config.  The server feature will be specfied via the
+`server_features` field described in [gRFC A30][A30].  The feature will
+be the string `trusted_xds_server`.  (Note that the name is intentionally
+fairly general, since it may be used to trigger other security-sensitive
+functionality in the future.)
+
+### xDS Resource Validation
+
+When validating an RDS resource, if the `trusted_xds_server`
+server option is present for this xDS server in the
+bootstrap config, we will look at the [`RouteAction.auto_host_rewrite`
+field](https://github.com/envoyproxy/envoy/blob/b65de1f56850326e1c6b74aa72cb1c9777441065/api/envoy/config/route/v3/route_components.proto#L1173)
+field.  The boolean value of this field will be included in the parsed
+resource struct that is passed to the XdsClient watcher.
+
+When validating an EDS resource, we will look at the [`Endpoint.hostname`
+field](https://github.com/envoyproxy/envoy/blob/b65de1f56850326e1c6b74aa72cb1c9777441065/api/envoy/config/endpoint/v3/endpoint_components.proto#L89).
+The string value of this field will be included in a per-endpoint resolver
+`hostname` attribute in the parsed resource struct that is passed to the
+XdsClient watcher.  Note that the resolver attribute used here should
+be a general-purpose one, not something specific to EDS; for example, in
+the future, we may want to use this same attribute to expose per-endpoint
```

Comment:
What is this about? If we're going to be looking at this for general things, I don't really understand the semantics. Right now this is happening independent of `trusted_xds_server` (is that right?) and `auto_host_rewrite`, so this is just a random string at this point from the perspective of the system. I'm very wary calling this attribute "hostname," even when trusted_xds_server=true, because it is conditional on other things. If it were only for xds, maybe I could live with it, but if we are having it be for random other things too... we need to very clearly specify what the value means/is. I have no idea what semantics the value has when trusted_xds_server=false.

Reply:
The idea is that this is a general-purpose attribute (not xDS-specific) to indicate a human-readable name associated with the individual address.  Any resolver can populate this attribute, and if it's set, we can display this information in channelz.

This is something we started working on a long time ago for internal purposes but never really drove to completion.  In C-core, we added the attribute in https://github.com/grpc/grpc/pull/24542, and a bunch of internal work was done to include this info in our internal resolver in b/130755297 and b/130754689, but the remaining work of actually plumbing it into channelz hasn't been done yet (b/130755264).

Reply:
There's still some awkwardness and the semantics are a bit strange (generally I'd write, "this value is untrusted; do not use it for security" but here we're immediately going to use it for security). But that's all fine; we don't paint ourselves into any corner here. We can "do something" here and see where things go.